### PR TITLE
Add bookmark feature

### LINKED ISSUE #71: Add Bookmark functionality
Add bookmark feature which includes:
- BookmarkCommand to bookmark a company
- RemoveBookmarkCommand to remove a bookmarked company
- BookmarkListCommand to show a list of bookmarked companies

### PR DESCRIPTION
**Pull request by vinceg4**
Add bookmark feature which includes:
- BookmarkCommand to bookmark a company
- RemoveBookmarkCommand to remove a bookmarked company
- BookmarkListCommand to list all the bookmarked companies

Test cases have been written for the above components as well.

Closes #71

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/BookmarkListCommand.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public static final String MESSAGE_NO_BOOKMARKED_COMPANIES = "There is no bookmarked companies";
+
+    public static final Predicate<Company> PREDICATE_BOOKMARKED_COMPANIES = c -> c.getIsBookmark().getIsBookmarkValue();
+
+    @Override
+    public CommandResult execute(Model model) {
+        requireNonNull(model);
+
+        model.updateFilteredCompanyList(PREDICATE_BOOKMARKED_COMPANIES);
+        List<?> filteredCompanies = model.getFilteredCompanyList();
```
Comment (LingZiyann) [inline src/main/java/seedu/address/logic/commands/BookmarkListCommand.java:29]: Could the wildcard "?" cause any issue? Maybe better to use List< company > ?

**Review by LingZiyann — APPROVED** [review]
Looks good in general. Although for writing tests, perhaps you could 'structure' your logic together that makes it easier to read by adding comments ? For example, you can group the code to 3 sections using comments
    // Arrange: Set up initial conditions and prepare data. Creating your necessary instances
    // Act: Perform the action under test. Execute whatever methods you need to
    // Assert: assert statements used here

File: src/main/java/seedu/address/logic/commands/BookmarkListCommand.java (hunk 1 above)
Comment (vinceg4) [reply src/main/java/seedu/address/logic/commands/BookmarkListCommand.java]: Have changed it to List<Company> 👍🏻

**Comment by LingZiyann** [comment]
LGTM! Btw the comments thing is just a suggestion, you can choose whether to write it or not

File: src/test/java/seedu/address/logic/commands/BookmarkCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import seedu.address.commons.core.index.Index;
+import seedu.address.logic.Messages;
+import seedu.address.model.Model;
+import seedu.address.model.ModelManager;
+import seedu.address.model.UserPrefs;
+import seedu.address.model.company.Bookmark;
+import seedu.address.model.company.Company;
+
+
+public class BookmarkCommandTest {
```
Comment (mattisongjj) [inline src/test/java/seedu/address/logic/commands/BookmarkCommandTest.java:24]: Test cases are extensive and covers each case well, (filtered and unfiltered list are both considered) and invalid indexes and existing bookmark cases are also well covered!

File: src/main/java/seedu/address/logic/commands/BookmarkCommand.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    public BookmarkCommand(Index index) {
+        requireAllNonNull(index);
+        this.index = index;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        List<Company> lastShownList = model.getFilteredCompanyList();
+
+        if (index.getZeroBased() >= lastShownList.size()) {
```
Comment (mattisongjj) [inline src/main/java/seedu/address/logic/commands/BookmarkCommand.java:46]: Checking for index out of bounds condition is good here, shows good defensive programming

File: src/main/java/seedu/address/logic/commands/BookmarkCommand.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        model.updateFilteredCompanyList(Model.PREDICATE_SHOW_ALL_COMPANIES);
+
+        return new CommandResult(String.format(MESSAGE_BOOKMARK_SUCCESS,
+                Messages.format(companyToBookmark)));
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
```
Comment (mattisongjj) [inline src/main/java/seedu/address/logic/commands/BookmarkCommand.java:72]: Could add a test case for this line to check that the same BookMark instance always returns true when compared

File: src/main/java/seedu/address/logic/commands/BookmarkCommand.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        if (!(other instanceof BookmarkCommand)) {
+            return false;
```
Comment (mattisongjj) [inline src/main/java/seedu/address/logic/commands/BookmarkCommand.java:76]: Another test case similar to the one above could be added for this line, checking that Bookmark commands do not return True when compared with other objects

File: src/main/java/seedu/address/logic/commands/RemoveBookmarkCommand.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        if (!(other instanceof RemoveBookmarkCommand)) {
+            return false;
```
Comment (mattisongjj) [inline src/main/java/seedu/address/logic/commands/RemoveBookmarkCommand.java:76]: More test cases could be added for the equals object. (See my comments for BoomarkCommand)

**Review by mattisongjj — COMMENTED** [review]
Overall good, just some nitpicking

**Pull request merged**